### ISSUE TITLE
Tracking issue for libsyntax breaking changes

### ISSUE THREAD

**Issue by Manishearth** [rust-lang/rust#31645]
See also: https://internals.rust-lang.org/t/please-consider-stability-of-libsyntax/2947

We decided to batch up breaking changes to oft-used unstable compiler internals (mainly, libsyntax) until they're large enough; so that the impact of these breaking changes is minimized. (For example, every time libsyntax breaks, aster and transitively everyone using serde on nightly has to fix it)

Let's try to coordinate these in the future through this issue. We can list breaking PRs here and batch them up when we feel that they're large enough (or when there's another inevitable breaking change about to merge). Merge them when there's enough time for the folks below to know about it, and preferably make PRs to the larger projects below beforehand.

CCing stakeholders:
- libsyntax: @nrc @huonw @eddyb
- clippy: me, @llogiq
- diesel: @sgrif
- phf: @sfackler
- regex_macros: @BurntSushi, @alexcrichton 
- rust-cpp, other plugins: @mystor

Add more if you know of them!

**Comment by BurntSushi** [comment]
I would also like to re-submit my plea that commits with breaking changes are tagged, and if possible, at least some message that hints at or explains how to fix breakages.

Other crates with macros in use (apparently) are `docopt_macros` and `quickcheck_macros`.

**Comment by Manishearth** [comment]
Yeah, they get tagged with [breaking-change]. With this bug in place they should also be posted here first and we can notify everyone well in advance before a rollup.

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/pull/32767 is a breaking change that will land soon, along with #32688.

I'd normally wait a few days before merging these so that everyone can fix their code early, but in this case #32688 already landed so there's not much I can do.


This will definitely break syntex and friends, cc @erickt. I've never looked at diesel's code so it may or may not break it, @sgrif you may want to have a look before this lands in the nightly.

**Comment by sgrif** [comment]
Thanks for the ping, definitely going to cause breakage.

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/pull/33041 will land in a few days. It changes the parser and [some of the token tree representation](https://github.com/rust-lang/rust/pull/33041/files#diff-74b5ba63896c56c0964426955d2700d1L153)

 - Definitely will break syntex/aster/serde (cc @erickt)
 - Might break Diesel, Regex, phf (cc @sgrif, @sfackler, @BurntSushi): anyone who uses macros operating on token trees.

**Comment by Manishearth** [comment]
Note that https://github.com/rust-lang/rust/pull/31414 has been added to the rollup. It will break derive plugins, not sure what else.

**Comment by Manishearth** [comment]
Also https://github.com/rust-lang/rust/pull/33125

So far, we will be landing
 
 - https://github.com/rust-lang/rust/pull/33125
 -  #31414 
 - #33041 

once they all get approved

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/pull/33157 is also in the list

**Comment by Manishearth** [comment]
Merging now: https://github.com/rust-lang/rust/pull/33179

**Comment by Manishearth** [comment]
It landed

**Comment by Manishearth** [comment]
For next batch (probably won't be any time soon):

 - https://github.com/rust-lang/rust/pull/33351

**Comment by jimmycuadra** [comment]
I don't think @erickt works on Serde very actively at this point, since he has lots of other duties. Is there another maintainer that should be CCed on this issue?

**Comment by Manishearth** [comment]
I was going to start doing serde ups but I haven't had the time and I might not get it for a month.

**Comment by Manishearth** [comment]
@petrochenkov Probably should make your PRs nowish given that the range one will need to land soon.

**Comment by Manishearth** [comment]
Batch should happen in a day or two: https://github.com/rust-lang/rust/pull/33864

cc @erickt @dtolnay @sgrif

**Comment by Manishearth** [comment]
Batch landed!

**Comment by kamalmarhubi** [comment]
@Manishearth 
> I was going to start doing serde ups but I haven't had the time and I might not get it for a month.

What's involved? I'm interested in keeping `syntex_syntax` close to `libsyntax` so that rustfmt keeps up.

**Comment by Manishearth** [comment]
Um, ask @erickt

**Comment by Manishearth** [comment]
@kamalmarhubi https://github.com/serde-rs/syntex/blob/master/CONTRIBUTING.md#how-to-do-a-syntex_syntax-release

and then you fix up aster and quasi to do the same

**Comment by kamalmarhubi** [comment]
@Manishearth I was about to come here and link that to answer myself, but you're quick! :-)

**Comment by Manishearth** [comment]
PRs for next batch:
 - https://github.com/rust-lang/rust/pull/34213 (affects aster, maybe: cc @erickt @dtolnay )

**Comment by dtolnay** [comment]
In case it affects anyone else - #34272 was a breaking change for Aster. https://github.com/serde-rs/aster/issues/88. I guess nobody noticed because rustup has been serving up nightly 2016-06-15 until recently.

This would be a good time to merge the other 5 if they are ready.

**Comment by mcarton** [comment]
Cc me.

(sorry for the noise but for some reason GitHub never notifies me for this despite I've subscribe to the thread)

**Comment by jseyfried** [comment]
@dtolnay sorry about that, I underestimated how much of the `libsyntax` API was being used by third parties. I'll be sure to check aster/serde in the future before landing anything borderline.

**Comment by Manishearth** [comment]
- #35591 (r=jntrnr)

**Comment by Manishearth** [comment]
Next breaking batch will contain the above PRs, waiting on the last one. cc @dtolnay @sfackler @mystor @sgrif @BurntSushi

**Comment by Manishearth** [comment]
Everything seems reviewed, breaking batch will be made todayish (and land over the weekend).

cc @dtolnay @BurntSushi @sgrif @sfackler @llogiq

**Comment by Manishearth** [comment]
Batch at https://github.com/rust-lang/rust/pull/36066

**Comment by Manishearth** [comment]
Landed!

**Comment by Manishearth** [comment]
Stuff for next batch:

 - https://github.com/rust-lang/rust/pull/35874

**Comment by mcarton** [comment]
@Manishearth don't know if it's to late for the two other PRs to make it into the next nightly, but https://github.com/rust-lang/rust/pull/36551 was a syntax breaking change and has already been merged.

**Comment by Manishearth** [comment]
Too late :/ But we're rolling it up now anyway

**Comment by Manishearth** [comment]
Breaking batch at https://github.com/rust-lang/rust/pull/36857 cc @dtolnay @sgrif @BurntSushi

**Comment by dtolnay** [comment]
This is exciting - I am not planning to do a serde_macros release so we will see if people revolt or move to serde_derive.

**Comment by Manishearth** [comment]
Should we PSA this?

**Comment by dtolnay** [comment]
The announcement that we are dropping serde_macros was at the top of reddit for a while and also popular on u.r-l.o. I think people will put two and two together when their builds fail. Worst case we get a handful of issues filed and we point them to the announcement. Do you know a better way to reach people who have not seen either of those already?

**Comment by sfackler** [comment]
You could potentially cut a new release that panics with a message telling people to move to serde-derive.

**Comment by mcarton** [comment]
@dtolnay Has this been announced in the Project Updates section of TWiR?

**Comment by dtolnay** [comment]
No we just missed the last one.

**Comment by Manishearth** [comment]
The panicky method feels a bit invasive imo.

It should be worth mentioning somewhere (perhaps on the crates.io page, or in the readme) that the method for upgrading is to switch over to the new macro version.

**Comment by eddyb** [comment]
If it has access to the diagnostics machinery, printing a deprecation warning wouldn't be that hard.

**Comment by Manishearth** [comment]
This landed. beware the breakage!

**Comment by sgrif** [comment]
We should be done with the macros 1.1 port later today so I no longer need to get pinged on these. Thank you for including me though!

**Comment by petrochenkov** [comment]
Proc macros are not tied to libsyntax now, so this is no longer relevant.
